### PR TITLE
refactor: :recycle: refactor workspace react

### PR DESCRIPTION
**Pull request by Gustav-Eikaas**
# Description

Features:
Wrapped tabs in error boundary and suspense
Fixed a lot of HMR errors
Clear query params on unmount

An ongoing process of fixing HMR...

Status:
HMR no longer breaks the application but adding/removing tabs or part of config is currently not supported. Only changes to data fetching will work...

Addresses:
[AB#35239](https://statoil-proview.visualstudio.com/9949ad5e-7d15-4531-901e-296574079ee1/_workitems/edit/35239)
[AB#34505](https://statoil-proview.visualstudio.com/9949ad5e-7d15-4531-901e-296574079ee1/_workitems/edit/34505)
[AB#35240](https://statoil-proview.visualstudio.com/9949ad5e-7d15-4531-901e-296574079ee1/_workitems/edit/35240)
[AB#35241](https://statoil-proview.visualstudio.com/9949ad5e-7d15-4531-901e-296574079ee1/_workitems/edit/35241)

Part of #263 #266 

## Checklist

-   [ ] My code follows the style guidelines of this project
-   [ ] Documentation and comments is added where needed.
-   [ ] My code is covered by tests.
-   [ ] UX has reviewed relevant UI contributions.

## Definition of Done
- [x] PR title and description are to the point and understandable
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: packages/workspace-fusion/src/lib/components/Workspace.tsx [hunk 1]
```diff
@@ -37,22 +34,42 @@ export function Workspace<
...
+				/>
 			</DataSourceProvider>
 		</QueryClientProvider>
 	);
 }
+
+/**
+ * Cleans up all query params used by fusion workspace when unmounting
+ */
+export const useCleanupQueryParams = (mediator: FusionMediator<any, any, any>, history: BrowserHistory) =>
```
Comment (Gustav-Eikaas) [inline packages/workspace-fusion/src/lib/components/Workspace.tsx:67]: Move to fusionUrlHandler

File: packages/workspace-fusion/src/lib/components/TabNavigation/TabNavigation.tsx [hunk 2]
```diff
@@ -10,16 +10,17 @@ type TabNavigationProps = {
 };
 
 export function TabNavigation({ children }: TabNavigationProps) {
-	const { tabController } = useControllerContext();
+	const set = useSetActiveTab();
 	const activeTab = useActiveTab();
+	const tabs = useTabs();
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/components/TabNavigation/TabNavigation.tsx:15]: Jeg tror jeg ville ha laget en ekstra hook for disse tre hook-ene.

const {setActiveTab, activeTab, allTabs} = useTabContext()

const useTabContext = () => {
    const setActiveTab = useSetActiveTab();
    const activeTab = useActiveTab();
    const allTabs = useTabs();
    return {setActiveTab, activeTab, allTabs}
}

File: packages/workspace-fusion/src/lib/components/Workspace.tsx [hunk 3]
```diff
@@ -1,29 +1,26 @@
...
+import history from 'history/browser';
+import { BrowserHistory } from 'history';
 
 /** Only gets called once */
 const useStable = <T,>(val: T) => {
 	return useState(() => val)[0];
 };
 
-function useCheckParentClient(): QueryClient | undefined {
+function useCheckParentClient(): QueryClient {
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/components/Workspace.tsx:19]: Legg inn tsdoc for denne, ikke så lett å forstå uten

File: packages/workspace-fusion/src/lib/components/Workspace.tsx [hunk 4]
```diff
@@ -1,29 +1,26 @@
...
 import { DataSourceProvider } from '../integrations/data-source';
 import { QueryClient, QueryClientProvider, useQueryClient } from 'react-query';
+import { fusionQueryParams, updateQueryParams } from '../classes/fusionUrlHandler';
+import history from 'history/browser';
+import { BrowserHistory } from 'history';
 
 /** Only gets called once */
 const useStable = <T,>(val: T) => {
 	return useState(() => val)[0];
 };
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/components/Workspace.tsx:17]: Denne hooken og useCheckParentClient hadde jeg lagt i en egen hooks mappe

Comment (jaysencpp) [reply packages/workspace-fusion/src/lib/components/Workspace.tsx:17]: useCleanupParams hooken på bunnen også

File: packages/workspace-fusion/src/lib/integrations/common/components/TabNavigation.tsx [hunk 5]
```diff
@@ -7,7 +7,7 @@ import { Fragment } from 'react';
 export const TabNavigation = () => {
 	const { icons, analyticsTabs, viewTabs } = useWorkspaceHeaderComponents();
 
-	const { tabController } = useControllerContext<WorkspaceTabNames, unknown>();
+	const setActiveTab = useSetActiveTab();
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/common/components/TabNavigation.tsx:10]: Vurder useTabContext()

File: packages/workspace-fusion/src/lib/integrations/data-source/components/DataLoader.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+};
+
+const ErrorComponent = ({ error, resetErrorBoundary }: FallbackProps) => (
+	<StyledCentering>
+		<div style={{ display: 'flex', flexDirection: 'column', gap: '0.5em' }}>
+			<div>Failed to load data!</div>
+			<Button variant="outlined" onClick={() => resetErrorBoundary()}>
+				Retry
+			</Button>
+			<pre>{JSON.stringify(error)}</pre>
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/data-source/components/DataLoader.tsx:34]: Hadde lagt på for json error, tror det er noe sånn JSON.stringify(error, undefined, 4)

File: packages/workspace-fusion/src/lib/integrations/filter/utils/addFilterContext.tsx [hunk 7]
```diff
@@ -1,36 +1,61 @@
...
 
 /**
  * Wraps workspace in filter context
  */
-export function addFilterContext<TData, TError>(
-	viewController: WorkspaceViewController<WorkspaceTabNames, TError>,
-	filterController: ReactFilterController<TData>
+export function makeFilterProvider<TData, TError>(
+	filterController: ReactFilterController<TData>,
+	mediator: FusionMediator<any, any, any>
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/filter/utils/addFilterContext.tsx:14]: mediator typing?

File: packages/workspace-fusion/src/lib/integrations/filter/utils/addFilterContext.tsx [hunk 8]
```diff
@@ -1,36 +1,61 @@
...
-		onSuccess(data) {
-			filterControlLer.setData(data as unknown[]);
-			filterControlLer.init();
-		},
+		useErrorBoundary: false,
+		suspense: false,
 	});
+
+	useEffect(() => {
+		filterControlLer.setData(data as unknown[]);
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/filter/utils/addFilterContext.tsx:37]: Data kan vel være undefined her? Før så satt du data onSuccess, men nå settes data før API kallet er ferdig?

File: packages/workspace-fusion/src/lib/integrations/data-source/components/DataLoader.tsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import { Button } from '@equinor/eds-core-react';
+import { PropsWithChildren } from 'react';
+import { ErrorBoundary, FallbackProps } from 'react-error-boundary';
+import { QueryErrorResetBoundary, useQuery } from 'react-query';
+import styled from 'styled-components';
+import { useQueryContext } from '../context';
+
+function Query({ children }: PropsWithChildren) {
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/data-source/components/DataLoader.tsx:8]: Trenger en forklaring her. Hva er poenget med denne?

File: packages/workspace-fusion/src/lib/integrations/grid/components/addGrid.tsx [hunk 10]
```diff
@@ -1,42 +1,62 @@
...
+import { Provider } from '@equinor/workspace-react';
+import { createGridController, GridController } from '@equinor/workspace-ag-grid';
 import { GridIcon } from '../icons/GridIcon';
-import { FusionMediator, GetIdentifier, WorkspaceTabNames } from '../../../types';
-import { configureBookmark } from '../utils/configureBookmark';
-import { configureDataChange } from '../utils/configureDataChange';
-import { configureHighlightSelection } from '../utils/configureHighlightSelection';
+import { FusionMediator } from '../../../types';
+import { useBookmarkService } from '../utils/configureBookmark';
+import { configureDataChange as useDataChange } from '../utils/configureDataChange';
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/grid/components/addGrid.tsx:6]: Klarer ikke å se hvor configureDataChange kommer fra. configureDataChange har bare én named export som er useOnDataChanged?

File: packages/workspace-fusion/src/lib/integrations/grid/components/addGrid.tsx [hunk 11]
```diff
@@ -1,42 +1,62 @@
...
+				</DataLoader>
+			),
+			name: 'grid',
+			TabIcon: GridIcon,
+			CustomHeader: () => <GridHeader controller={gridController} />,
+		},
+	};
+}
+
+function useContextService(mediator: FusionMediator<any, any, any>, gridController: GridController<any, any>) {
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/grid/components/addGrid.tsx:53]: Denne hooken passer sammen med resten som er i utils mappa

Comment (jaysencpp) [reply packages/workspace-fusion/src/lib/integrations/grid/components/addGrid.tsx:53]: Trenger en forklaring på hvorfor dette skal være en hook :D

File: packages/workspace-fusion/src/lib/integrations/sidesheet/components/addSidesheet.tsx [hunk 12]
```diff
@@ -12,25 +13,46 @@ export function addSidesheet<
...
+				<SidesheetSimpleWrapper config={config} mediator={mediator} />
+			</Wrapper>
+		);
 	}
+	return;
+}
+
+type Props = {
+	children: ReactNode;
+	mediator: FusionMediator<any, any, any>;
```
Comment (jaysencpp) [inline packages/workspace-fusion/src/lib/integrations/sidesheet/components/addSidesheet.tsx:39]: mediator typing?

File: packages/workspace-fusion/src/lib/components/TabNavigation/TabNavigation.tsx (hunk 2 above)
Comment (jaysencpp) [reply packages/workspace-fusion/src/lib/components/TabNavigation/TabNavigation.tsx:15]: Jeg har lest gjennom useTabsContext og zustand koden nå, det gir mer mening nå at det splitta opp

File: packages/workspace-fusion/src/lib/integrations/filter/utils/addFilterContext.tsx (hunk 7 above)
Comment (Gustav-Eikaas) [reply packages/workspace-fusion/src/lib/integrations/filter/utils/addFilterContext.tsx:14]: So many context's that I honestly dont see the point anymore. Unless the end-user config has any typings

File: packages/workspace-fusion/src/lib/integrations/garden/components/addGarden.tsx [hunk 13]
```diff
@@ -19,30 +21,41 @@ export function addGarden<
...
 
-	viewController.tabController.addTab({
-		Component: () => <GardenWrapper controller={gardenController} mediator={mediator} />,
-		name: 'garden',
-		TabIcon: GardenIcon,
-		CustomHeader: () => <GardenWorkspaceHeader controller={gardenController} />,
-	});
+	const provider: Provider = {
+		Component: ({ children }) => {
+			useEffect(useOnDataChanged(gardenController, mediator), [mediator]);
```
Comment (Noggling) [inline packages/workspace-fusion/src/lib/integrations/garden/components/addGarden.tsx:39]: Rename to dataChangedEffect

Comment (Noggling) [reply packages/workspace-fusion/src/lib/integrations/garden/components/addGarden.tsx:39]: -->

File: packages/workspace-fusion/src/lib/integrations/data-source/components/DataLoader.tsx [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+			{({ reset }) => (
+				<ErrorBoundary onReset={reset} fallbackRender={ErrorComponent}>
+					<Query>{children}</Query>
+				</ErrorBoundary>
+			)}
+		</QueryErrorResetBoundary>
+	);
+};
+
+const ErrorComponent = ({ error, resetErrorBoundary }: FallbackProps) => (
```
Comment (Noggling) [inline packages/workspace-fusion/src/lib/integrations/data-source/components/DataLoader.tsx:30]: Fix

File: packages/workspace-react/src/lib/components/Workspace.tsx [hunk 15]
```diff
@@ -1,24 +1,95 @@
...
+	defaultTab?: string;
+	tabs: Tab[];
+	Sidesheet?: () => JSX.Element;
+	providers: Provider[];
+	events?: WorkspaceEvents;
 }
 
-export function Workspace<TTabNames extends string, TError>({ controller }: WorkspaceProps<TTabNames, TError>) {
-	useSetFallbackTab(controller);
+const TabProvider = createContext<StoreApi<TabController> | null>(null);
```
Comment (Noggling) [inline packages/workspace-react/src/lib/components/Workspace.tsx:21]: Flytte ut?

File: packages/workspace-react/src/lib/components/Workspace.tsx [hunk 16]
```diff
@@ -1,24 +1,60 @@
...
+						<WorkspaceBody>
+							<Sidesheet />
+						</WorkspaceBody>
+					</ContextProviders>
+				</EventHandler>
+			</TabProvider.Provider>
 		</WorkspaceWrapper>
 	);
 }
+
```
Comment (Noggling) [inline packages/workspace-react/src/lib/components/Workspace.tsx:50]: Splitt fil

**Review by Noggling — APPROVED** [review]

**Pull request merged**